src/Features/LanguageServer/Protocol/Handler/Symbols/DocumentSymbolsHandler.cs

```diff
@@ -0,0 +1,201 @@
+            }
+
+            return Create(item, location.SourceSpan, containerName, document, text);
+
+            static SymbolInformation Create(NavigationBarItem item, TextSpan span, string containerName, Document document, SourceText text)
+            {
+                return new SymbolInformation
+                {
+                    Name = item.Text,
+                    Location = new LSP.Location
+                    {
+                        Uri = document.GetURI(),
+                        Range = ProtocolConversions.TextSpanToRange(span, text),
+                    },
+                    Kind = ProtocolConversions.GlyphToSymbolKind(item.Glyph),
+                    ContainerName = containerName,
+                };
+            }
+        }
+
+        /// <summary>
+        /// Get a document symbol from a specified nav bar item.
+        /// </summary>
+        private static async Task<DocumentSymbol> GetDocumentSymbolAsync(NavigationBarItem item, Compilation compilation, SyntaxTree tree,
+            SourceText text, CancellationToken cancellationToken)
+        {
+            // it is actually symbol location getter. but anyway.
+            var location = GetLocation(item, compilation, tree, cancellationToken);
+            if (location == null)
+            {
+                return null;
+            }
+
+            var symbol = await GetSymbolAsync(location, compilation, cancellationToken).ConfigureAwait(false);
+            if (symbol == null)
+            {
+                return null;
+            }
+
+            return new DocumentSymbol
+            {
+                Name = symbol.Name,
+                Detail = item.Text,
+                Kind = ProtocolConversions.GlyphToSymbolKind(item.Glyph),
+                Deprecated = symbol.GetAttributes().Any(x => x.AttributeClass.MetadataName == "ObsoleteAttribute"),
+                Range = ProtocolConversions.TextSpanToRange(item.Spans.First(), text),
+                SelectionRange = ProtocolConversions.TextSpanToRange(location.SourceSpan, text),
+                Children = await GetChildrenAsync(item.ChildItems, compilation, tree, text, cancellationToken).ConfigureAwait(false),
+            };
+
+            static async Task<DocumentSymbol[]> GetChildrenAsync(IEnumerable<NavigationBarItem> items, Compilation compilation, SyntaxTree tree,
+                SourceText text, CancellationToken cancellationToken)
+            {
+                var list = new ArrayBuilder<DocumentSymbol>();
+                foreach (var item in items)
+                {
+                    list.Add(await GetDocumentSymbolAsync(item, compilation, tree, text, cancellationToken).ConfigureAwait(false));
+                }
+
+                return list.ToArrayAndFree();
+            }
+
+            static async Task<ISymbol> GetSymbolAsync(Location location, Compilation compilation, CancellationToken cancellationToken)
```

Comment:
>GetSymbolAsync [](start = 39, length = 14)

I'm curious why do we have this logic here. My understanding is that this code is only used by LSP. Do we have different behavior for regular VS vs LSP? That seems wrong.

Reply:
Note - this is the liveshare implementation, it may not match exactly and we will be coming back to change a lot of stuff.  Just want to get it merged so we can avoid having two copies of the code.

Is there a standard method I should use to retrieve the symbol at a location?  I found a bunch of different implementations from a cursory search.

Reply:
Just to speak to why I'm hesitant to make changes here, I want to make the insertion as simple as possible. Since we're copying everything over from liveshare, there's a lot to test and a lot to go wrong.  In this PR I'd rather avoid changing the implementation as much as possible so we can do 1 step at a time.  I can add an item to the project to come back to this once liveshare is using us, and test the changes here in isolation (which is what my plan has been so far).

Does that sound reasonable @tmat ?

Reply:
Filed - https://github.com/dotnet/roslyn/projects/45#card-23220339

Reply:
I see. Sounds good. Yes, let's follow up later.